### ISSUE TITLE
OpenCV 4.7.0 Huawei CANN run Yunet at very low speed: AI core and Ctrl cpu very Low

### ISSUE THREAD

**Issue by hz6888** [opencv/opencv#23281]
### System Information

OpenCV version: 4.7.0
Operating System / Platform: Ubuntu 18.04
Compiler & compiler version: GCC 7.5.0

### Detailed description

Refering back to previous issue: 
https://github.com/opencv/opencv/issues/23199     OpenCV 4.7.0 Huawei CANN support error

The code could work with "Ascend310P3" change. But I checked the speed of the faceDetectorYN, i.e. Yunet, found the AI core and Ctrl cpu are very low. 

NpuID(Idx)  ChipId(Idx) Pwr(W)      Temp(C)     AI Core(%)  AI Cpu(%)   Ctrl Cpu(%) Memory(%)   Memory BW(%)
2           0           NA          44          0           0           26          5           2
3           0           NA          46          0           0           0           4           51
2           0           NA          44          6           0           0           5           2
3           0           NA          46          0           0           2           4           51
2           0           NA          45          6           0           0           5           2
3           0           NA          46          0           0           1           4           51
2           0           NA          45          0           0           1           5           2
3           0           NA          46          0           0           1           4           51
2           0           NA          45          6           0           2           5           2
3           0           NA          46          0           0           1           4           51
2           0           NA          45          6           0           0           5           2
3           0           NA          46          0           0           1           4           51
2           0           NA          46          0           0           2           5           2
3           0           NA          46          0           0           0           4           51
2           0           NA          46          0           0           1           5           2
3           0           NA          46          0           0           1           4           51
2           0           NA          46          6           0           0           5           2
3           0           NA          46          0           0           3           4           51

I compare to samples of ppresnet50, mobilev1, yolox, the results are like:

NpuID(Idx)  ChipId(Idx) Pwr(W)      Temp(C)     AI Core(%)  AI Cpu(%)   Ctrl Cpu(%) Memory(%)   Memory BW(%)
3           0           NA          46          0           0           1           4           51
2           0           NA          43          0           0           2           4           51
3           0           NA          46          0           0           1           4           51
2           0           NA          43          0           0           24          4           19
3           0           NA          46          0           0           1           4           51
2           0           NA          46          37          0           1           4           0
3           0           NA          46          0           0           1           4           51
2           0           NA          45          0           0           16          4           0
3           0           NA          46          0           0           2           4           51
2           0           NA          45          0           0           1           4           0
3           0           NA          46          0           0           2           4           51

NpuID(Idx)  ChipId(Idx) Pwr(W)      Temp(C)     AI Core(%)  AI Cpu(%)   Ctrl Cpu(%) Memory(%)   Memory BW(%)
2           0           NA          50          0           0           1           4           0
3           0           NA          46          0           0           2           4           51
2           0           NA          50          0           0           1           4           51
3           0           NA          46          0           0           2           4           51
2           0           NA          50          0           0           0           4           51
3           0           NA          46          0           0           2           4           51
2           0           NA          50          12          0           26          4           0
3           0           NA          46          0           0           1           4           51
2           0           NA          50          0           0           21          4           0
3           0           NA          46          0           0           1           4           51
2           0           NA          50          0           0           2           4           0
3           0           NA          46          0           0           1           4           51

NpuID(Idx)  ChipId(Idx) Pwr(W)      Temp(C)     AI Core(%)  AI Cpu(%)   Ctrl Cpu(%) Memory(%)   Memory BW(%)
3           0           NA          46          0           0           1           4           51
2           0           NA          46          0           0           17          4           0
3           0           NA          46          0           0           0           4           51
2           0           NA          48          29          0           8           4           3
3           0           NA          46          0           0           1           4           51
2           0           NA          49          26          0           2           4           3
3           0           NA          46          0           0           0           4           51
2           0           NA          49          28          0           1           4           3
3           0           NA          46          0           0           1           4           51
2           0           NA          49          29          0           1           4           3
3           0           NA          46          0           0           1           4           51
2           0           NA          50          29          0           0           4           3
3           0           NA          46          0           0           0           4           51
2           0           NA          50          19          0           40          4           0
3           0           NA          46          0           0           2           4           51
2           0           NA          49          0           0           7           4           0
3           0           NA          46          0           0           1           4           51
2           0           NA          49          0           0           0           4           0

I'm not sure about these results ?  The hardware and software are working fine ?

Or the "Ascend310P3" SoC version is still not really OK in OpenCV 4.7.0 ?

### Steps to reproduce

npu-smi info watch

Any better tool to analyse the performance ?

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by fengyuentau** [comment]
YuNet is quite lightweight (~300KB) compared with ResNet50 (~100MB) and YOLOX (~34MB). So running YuNet with single batch on the CANN backend does not put a lot of loads on the NPU. One more possible reason of the low speed can be the images having different sizes. Current implementation of CANN backend assumes a fixed input size, and the graph is compiled for this fixed input size. Once it is changed, the graph will be compiled again, leading to extremely low inference speed.

So possible solutions are

- Use big batch to fully utilize multiple cores;
- Resize all images to a single size;
- Compare with original CANN APIs. If performance is still low using big batch & original CANN APIs, it is probably something related to the CANN library, and you might go [gitee.com/cann](gitee.com/cann) to report an issue.

**Comment by hz6888** [comment]
YuNet is lightweight, but the infer speed is low, about 100ms-200ms per picture.

I use images of same size, actually I tried 1920x1080 and 480x270, both slow.

My test code is simply using OpenCV faceDetectorYN, i.e. dnn module, so it seems to have no "big batch" setting.  If I use native API of CANN, I may need to write preprocess and postprocess code, which is also a difficult job.


I found a YuNet solution, based on MindX 
  https://gitee.com/ascend/mindxsdk-referenceapps/tree/master/contrib/yunet
I'm not sure if it works, but it contains preprocess and postprecess code of YuNet.

**Comment by fengyuentau** [comment]
> My test code is simply using OpenCV faceDetectorYN, i.e. dnn module, so it seems to have no "big batch" setting.

Yes, it is designed without the consideration of batch. It can be a feature in the next release. By the way, what is the exact speed with 1920x1080 and 480x270 images as input?

**Comment by hz6888** [comment]
1920x1080
iterate: 810 times;  total time: 86.3618 s;  average time: 106.62 ms;    // from input to infer output
iterate: 810 times;  total time: 93.7088 s;  average time: 115.69 ms;    // from input to draw face keypoints rectangle circle
iterate: 810 times;  total time: 173.973 s;  average time: 214.781 ms;  // from input to imwrite output image

480x270
iterate: 810 times;  total time: 167.776 s;  average time: 207.131 ms;  // from input to imwrite output image


But the sample codes of ppresnet50, mobilev1, yolox, don't have batch either.   

When I test them by using 1000 iterations from input to output, the speed, AI core, Ctrl cpu are much higher than YuNet.

**Comment by fengyuentau** [comment]
> 1920x1080
> iterate: 810 times; total time: 173.973 s; average time: 214.781 ms; // from input to imwrite output image
> 480x270
> iterate: 810 times; total time: 167.776 s; average time: 207.131 ms; // from input to imwrite output image

It looks weird to me that the difference is not too much. I will have a look on my side with Ascend 310. But I still insist that the model (YuNet) is too lightweight to fully utilize AI cores.
